nixos/modules/security/permissions-wrappers/permissions-wrapper.c

```diff
@@ -0,0 +1,224 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <assert.h>
+#include <errno.h>
+
+// Make sure assertions are not compiled out, we use them to codify
+// invariants about this program and we want it to fail fast and
+// loudly if they are violated.
+#undef NDEBUG
+
+extern char **environ;
+
+// The SOURCE_PROG and WRAPPER_DIR macros are supplied at compile time
+// for a security reason: So they cannot be changed at runtime.
+static char * sourceProg = SOURCE_PROG;
+static char * wrapperDir = WRAPPER_DIR;
+
+// Make sure we have the WRAPPER_TYPE macro specified at compile
+// time...
```

Comment:
It also means increasing chances that something is unfetchable from the binary cache for a relatively natural system configuration.

As for the second question — I just ask whether you could just check for ability to change capability set instead of compiling the file differently.

Also, not setting any of the setuid/setcap options is a run-time error. If you really want to go this way, please make it a compile-time error at least.